### PR TITLE
Fetches & Renders Forms

### PR DESCRIPTION
**Pull request by HassanAbouelela**
Closes this [notion ticket](https://www.notion.so/pythondiscord/Fetch-a-form-and-render-it-on-the-forms-route-559363e97356447da7378acb97a540ed).

- Fetches a form from the backend, and renders the question as well as the form page.
- Updates model in accordance with backend.
- Lays foundation for input validation and submission.

### CODE REVIEW

File: src/api/forms.ts [hunk 1]
```diff
@@ -1,43 +1,35 @@
...
 import ApiClient from "./client";
 
 export enum FormFeatures {
     Discoverable = "DISCOVERABLE",
     RequiresLogin = "REQUIRES_LOGIN",
     Open = "OPEN",
     CollectEmail = "COLLECT_EMAIL",
-    DisableAntispam = "DISABLE_ANTISPAM"
+    DisableAntispam = "DISABLE_ANTISPAM",
+    WEBHOOK_ENABLED = "WEBHOOK_ENABLED"
```
Comment (ks129) [inline src/api/forms.ts:10]:
```suggestion
    WebhookEnabled = "WEBHOOK_ENABLED"
```

Comment (HassanAbouelela) [reply src/api/forms.ts:10]: Thanks, good catch.

File: src/components/InputTypes/Select.tsx [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                    <div className="select_options">
+                        { this.props.options.map((option, index) => <div key={index}><hr/><div id="option">{option}</div></div>) }
+                    </div>
+                </div>
+            </div>
+        );
+    }
+}
+
+export default Select;
```
Comment (ks129) [inline src/components/InputTypes/Select.tsx:60]: GH shows that this file doesn't have newline at end.

Comment (HassanAbouelela) [reply src/components/InputTypes/Select.tsx]: Thanks, fixed.

File: src/components/HeaderBar/index.tsx [hunk 3]
```diff
@@ -1,59 +1,110 @@
...
-            <Header1 css={headerImageStyles}/>
-            <Header2 css={headerImageStyles}/>
+            <div css={headerImageStyles}>
+                <Header1/>
+                <Header2/>
+            </div>
+
+            <div css={css`${headerTextStyles}; margin-bottom: 12.5%;`}>
+                <h1 className="title">{title}</h1>
+                <h1 className="description">{description}</h1>
```
Comment (jb3) [inline src/components/HeaderBar/index.tsx:100]: Should we render this if we don't have a description passed?

I think this should work:
```suggestion
                {description ? <h1 className="description">{description}</h1> : null}
```

File: src/components/InputTypes/Select.tsx [hunk 4]
```diff
@@ -0,0 +1,183 @@
+/** @jsx jsx */
```
Comment (jb3) [inline src/components/InputTypes/Select.tsx:1]: Selects can't be closed by mouse presses.

https://user-images.githubusercontent.com/20439493/104824384-113d0280-5849-11eb-88fc-b077d45ec4ea.mov

File: src/components/InputTypes/index.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import Select from "./Select";
+import ShortText from "./ShortText";
+import TextArea from "./TextArea";
+
+import React, { ChangeEvent } from "react";
+
+import { QuestionType } from "../../api/question";
+import { QuestionProp } from "../Question";
+
+const _require_options: Array<QuestionType> = [
```
Comment (jb3) [inline src/components/InputTypes/index.tsx:14]: Not sure the leading underscore is necessary here. It's only really relevant in languages where everything is exported.

```suggestion
const require_options: Array<QuestionType> = [
```

File: src/components/InputTypes/index.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+export default function create_input({ question, public_state }: QuestionProp, handler: (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void): JSX.Element | JSX.Element[] {
+    let result: JSX.Element | JSX.Element[];
+
+    // eslint-disable-next-line
+    // @ts-ignore
+    let options: string[] = question.data["options"];
+
+    // Catch input types that require options but don't have any
+    if ((options === undefined || typeof options !== "object") && _require_options.includes(question.type)) {
```
Comment (jb3) [inline src/components/InputTypes/index.tsx:29]: Similar to above:
```suggestion
    if ((options === undefined || typeof options !== "object") && require_options.includes(question.type)) {
```

File: src/components/InputTypes/index.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        result = <Range question_id={question.id} options={options} handler={handler}/>;
+        break;
+
+    case QuestionType.Code:
+        // TODO: Implement
+        result = <Code handler={handler}/>;
+        break;
+
+    default:
+        result = <TextArea handler={handler}/>;
```
Comment (jb3) [inline src/components/InputTypes/index.tsx:66]: I'm not sure what the standard is, but maybe it'd look clearer if all cases were embedded four inside the switch block?

File: src/components/Question.tsx [hunk 8]
```diff
@@ -0,0 +1,166 @@
+/** @jsx jsx */
+import { jsx, css } from "@emotion/react";
+import React, { ChangeEvent } from "react";
+
+import { Question, QuestionType } from "../api/question";
+import { selectable } from "../commonStyles";
+import create_input from "./InputTypes";
+
+const _skip_normal_state: Array<QuestionType> = [
```
Comment (jb3) [inline src/components/Question.tsx:9]: Same as above:
```suggestion
const skip_normal_state: Array<QuestionType> = [
```

File: src/components/Question.tsx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+class RenderedQuestion extends React.Component<QuestionProp> {
+    constructor(props: QuestionProp) {
+        super(props);
+        if (props.question.type === QuestionType.TextArea) {
+            this.handler = this.text_area_handler.bind(this);
+        } else {
+            this.handler = this.normal_handler.bind(this);
+        }
+
+        if (!_skip_normal_state.includes(props.question.type)) {
```
Comment (jb3) [inline src/components/Question.tsx:31]:
```suggestion
        if (!skip_normal_state.includes(props.question.type)) {
```

File: src/components/Question.tsx [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+            this._setState("value", "");
+        }
+    }
+
+    _setState(target: string, value: string | boolean | null, callback?:() => void): void {
```
Comment (jb3) [inline src/components/Question.tsx:36]: Nicer method name, underscores aren't so nice:
```suggestion
            this.setPublicState("value", "");
        }
    }

    setPublicState(target: string, value: string | boolean | null, callback?:() => void): void {
```

File: src/components/Question.tsx [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            if (event.target.parentElement) {
+                value = event.target.parentElement.innerText.trimEnd();
+            } else {
+                value = event.target.value;
+            }
+            break;
+
+        default:
+            target = "value";
+            value = event.target.value;
```
Comment (jb3) [inline src/components/Question.tsx:66]: Same as above, maybe we can indent these cases.

File: src/components/Question.tsx [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        this._setState(target, value);
```
Comment (jb3) [inline src/components/Question.tsx:69]:
```suggestion
        this.setPublicState(target, value);
```

File: src/components/Question.tsx [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+
+        // Toggle checkbox class
+        if (event.target.type == "checkbox" && event.target.parentElement !== null) {
+            event.target.parentElement.classList.toggle("unselected");
+            event.target.parentElement.classList.toggle("selected");
+        }
+    }
+
+    text_area_handler(event: ChangeEvent<HTMLTextAreaElement>): void {
+        this._setState("value", event.target.value);
```
Comment (jb3) [inline src/components/Question.tsx:79]:
```suggestion
        this.setPublicState("value", event.target.value);
```

File: src/components/Question.tsx [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+        if (this.props.public_state.size === 0) {
+            switch (this.props.question.type) {
+            case QuestionType.Checkbox:
+                if (typeof options === "string") {
+                    return;
+                }
+
+                options.forEach((option, index) => {
+                    this._setState(`${("000" + index).slice(-4)}. ${option}`, false);
```
Comment (jb3) [inline src/components/Question.tsx:94]:
```suggestion
                    this.setPublicState(`${("000" + index).slice(-4)}. ${option}`, false);
```

File: src/components/Question.tsx [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+                });
+                break;
+
+            case QuestionType.Range:
+            case QuestionType.Radio:
+            case QuestionType.Select:
+                this._setState("value", null);
```
Comment (jb3) [inline src/components/Question.tsx:101]:
```suggestion
                this.setPublicState("value", null);
```

File: src/components/Question.tsx [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    constructor(props: QuestionProp) {
+        super(props);
+        if (props.question.type === QuestionType.TextArea) {
+            this.handler = this.text_area_handler.bind(this);
+        } else {
+            this.handler = this.normal_handler.bind(this);
+        }
+
+        if (!_skip_normal_state.includes(props.question.type)) {
+            this._setState("value", "");
```
Comment (jb3) [inline src/components/Question.tsx:32]:
```suggestion
            this.setPublicState("value", "");
```

**Review by jb3 — CHANGES_REQUESTED** [review]
Fantastic work! CSS looks good, though I'm not the best at it. Some minor comments.

File: src/components/InputTypes/index.tsx (hunk 7 above)
Comment (HassanAbouelela) [reply src/components/InputTypes/index.tsx]: I would agree. This was originally done to comply with the default eslint indent rules, but I've gone ahead and updated them.

File: src/components/InputTypes/Select.tsx (hunk 4 above)
Comment (HassanAbouelela) [reply src/components/InputTypes/Select.tsx:1]: Resolved in b6ae67ff.

**Review by jb3 — APPROVED** [review]

**Pull request merged**